### DISCUSSION TITLE
Computing shift vector and transition intensity

### DISCUSSION THREAD

**Discussion by starofus10** [wannier-berri/wannier-berri#292] [Q&A] [answered]
Dear all

I hope this issue finds you well.
I would like to ask about performing shift vector and transition intensity calculation in Wannierberri.

I cited to this paper https://www.nature.com/articles/s41524-023-01026-3
I wonder how to find the data for the shift vector and transition intensity from the shift current calculation using Wannierberri.
I would like to know the contribution of shift vector and transition intensity to the shift current magnitude.

Could you give me a suggestion to obtain the data that I needed to compute shift vector and transition intensity?
I wonder that we need to make a modification in the code.

Your insights and clarifications on these questions would be immensely helpful for me to better understand the code and its calculations. Thank you for your time and assistance.

Regards 
Joshua

**Comment by irukoa** [comment, marked as answer]
Hi @starofus10 ,

I am guessing that you require an amplementation of Eqs. (1), (2) from the paper you cited, in which case you will need $r_{nm}^{a}(k)$ and $R^{a;b}_{nm}(k)$. The first quantity is already implemented in the code, while the second can be obtained by employing the relation 

$$R_{nm}^{a;b}(k)=\frac{I_{nm}^{abb}(k)}{|r_{nm}^{b}(k)|^2},$$

which can be found, in e.g., [[Cook2017](http://www.nature.com/articles/ncomms14176)] (above Eq. 10). I would suggest to use the implementation of $I^{abb}_{nm}(k)$ in 

https://github.com/wannier-berri/wannier-berri/blob/91e5b6983bd82f6017877271e222d4030cf928b4/wannierberri/calculators/dynamic.py#L305 

and the implementation of $r_{nm}^{a}(k)$ given by `A_H = data_K.A_H`. By using these you could define your own calculators for the integrand of Eqs. (1) and (2) in the paper you cited (following the lines of already implemented calculators) and proceed as you would for any other calculation. I believe that you do not need to make any modification to the existing code.

Bests,
Álvaro
Edit: equation formatting.

**Comment by stepan-tsirkin** [comment]
@starofus10 @irukoa 

Just to note that $r_{mn}$ is defined as the off-diagonal Berry connection (diagonal part is zero by definition)

![image](https://github.com/wannier-berri/wannier-berri/assets/22172977/eeb1c05c-ec27-43c9-bc03-f3243f06595a)

while `A_H = data_K.A_H` gives the full one (diagonal and off-diagonal)

**Comment by Zlli16** [comment]
Dear all,

I have encountered a similar issue. 
How is the shift vector calculated? Is the method "class wannierberri.calculators.dynamic.ShiftCurrentFormula(data_K, sc_eta, **parameters)" used for this purpose? 
And how should the "data_K" parameter be specifically set?

Thank you for your time and assistance.

Sincerely,

Zonglun

**Comment by stepan-tsirkin** [comment]
the `data_K` is a temporary object of class `Data_K`,  that stores the data for a shifted FFT grid (see [paper](https://www.nature.com/articles/s41524-021-00498-5) . Normally the run() function creates and those data_K objects, and passes them to calculators, and after all calculators have works, the data_K is destroyed. However, you may create the object yourself, see 

https://github.com/wannier-berri/wannier-berri/blob/5d8d6d7373edcc3fe97841ea4a1af2052e927e46/wannierberri/data_K.py#L75-L110

The .ShiftCurrentFormula implements (see [paper](https://journals.aps.org/prb/pdf/10.1103/PhysRevB.97.245143) the matrix element 

<img width="225" alt="{234BA611-243F-4010-82C1-B03DA9D62381}" src="https://github.com/user-attachments/assets/7cf18686-c9ab-415d-afcd-f2d2ef526e63">

While the Shift  vector, according to [paper](https://www.nature.com/articles/s41524-023-01026-3)

is 

<img width="560" alt="{41AF7F3B-07B4-4F30-B39A-A6725E91ED40}" src="https://github.com/user-attachments/assets/a49f1ffa-faad-45ca-a0f5-6b8d174a2687">

@irukoa , what are the properties of this quantity? is it gauge-covariant/invariant ? Above you suggesed to use 

$$R^{a;b}{nm}(k)=\frac{I^{abb}{nm}(k)}{|r_{nm}^{b}|^2},$$

But is it well-behaved when $r_{nm}^b$ approaches zero?

**Comment by irukoa** [comment]
The shift vector is a gauge covariant quantity with respect to gauge transformations mixing $N$ bands and gauge invariant with respect to band diagonal gauge transformations. See Ref. [IATS](https://doi.org/10.1103/PhysRevB.97.245143), Sec. II.A or Ref. [PMOM](https://doi.org/10.1103/PhysRevB.99.045121), App. A.2.

Regarding stability as $r_{nm}^b \to 0$, perhaps I should elaborate a bit more my previous response: the shift vector is a quantity related to the polar decomposition of the dipole matrix element $r_{nm}^b$. If this last quantity approaches zero, the phase $\phi_{nm}^b$, and thus $R_{nm}^{a;b}$ become undefined. In practice, we depart from knowledge of $I_{nm}^{abb}$ and $r_{nm}^b$, so we can always check wether $r_{nm}^b$ is zero (or close enough to zero) before employing the method I proposed. In summary if $r_{nm}^b \to 0$, the shift vector loses its meaning, since $r_{nm}^b \to 0$ implies that the contribution from that $\textbf{k}$-point to the shift current response $\sigma^{abb}(\omega)$ is zero.

**Comment by hamzaze-coder** [comment]
Dear @irukoa and @stepan-tsirkin,

I am currently working on reproducing the k-resolved shift current and shift vector
as shown in Fei, Tan & Rappe (PRB 2020), and I would like to compute the
quantities:

- the shift vector \( R^{a;b}_{nm}(k) \),
- and the shift-current integrand (before Brillouin-zone integration).

I have read your previous answers:

- @irukoa (Nov 17, 2023), where you explained that  
  \( R^{a;b}_{nm}(k) = I^{ab}_{nm}(k) / | r^b_{nm}(k) |^2 \),  
  and that the needed quantities are already implemented in  
  `wannierberri/calculators/dynamic.py`.

- @stepan-tsirkin (Oct 23, 2024), where you described how the `Data_K` object  
  contains all data computed on the FFT grid and can, in principle, be created  
  manually for low-level access.

I would like to respectfully ask a few questions:

---

### **1. Is there an existing Python example or calculator showing how to compute the shift vector explicitly?**
